nussl/separation/duet.py

```diff
@@ -0,0 +1,440 @@
+        copy_col = int(np.floor(kcol / 2))  # number of columns to copy on either side
+
+        # TODO: This is very ugly. Make this readable.
+        # form the augmented matrix (rows and columns added to top, bottom, and sides)
+        matrix = np.mat(matrix)  # make sure Mat is a Numpy matrix
+        augmented_matrix = np.vstack(
+            [
+                np.hstack(
+                    [matrix[0, 0] * np.ones((copy_row, copy_col)),
+                     np.ones((copy_row, 1)) * matrix[0, :],
+                     matrix[0, -1] * np.ones((copy_row, copy_col))
+                     ]),
+                np.hstack(
+                    [matrix[:, 0] * np.ones((1, copy_col)),
+                     matrix,
+                     matrix[:, -1] * np.ones((1, copy_col))]),
+                np.hstack(
+                    [matrix[-1, 1] * np.ones((copy_row, copy_col)),
+                     np.ones((copy_row, 1)) * matrix[-1, :],
+                     matrix[-1, -1] * np.ones((copy_row, copy_col))
+                     ]
+                )
+            ]
+        )
+
+        # perform two-dimensional convolution between the input matrix and the kernel
+        smooted_matrix = signal.convolve2d(augmented_matrix, kernel_matrix[::-1, ::-1], mode='valid')
+
+        return smooted_matrix
+
+    def make_audio_signals(self):
+        """Returns the extracted signals
+
+        Returns:
+            signals (List[AudioSignal]): List of AudioSignals extracted using DUET.
+
+        """
+        signals = []
+        for i in range(self.num_sources):
+            # Apply masks to stft channels using equation provided by Rickard
+            stft_sources = ((self.stft_ch0 + self.atn_peak[i] * np.exp(1j * self.frequency_matrix * self.delay_peak[i])
+                             * self.stft_ch1) / (1 + self.atn_peak[i] ** 2))
+            new_sig = self.audio_signal.make_copy_with_stft_data(stft_sources, verbose=False)
+            new_sig = new_sig.apply_mask(
+                self.masks[i])  # Ask Ethan if we can have a apply_mask that doesn't return a new mask!
+            new_sig.stft_params = self.stft_params
+            source_estimate = new_sig.istft(overwrite=True, truncate_to_length=self.audio_signal.signal_length)
+            cur_signal = nussl.audio_signal.AudioSignal(audio_data_array=source_estimate,
+                                                        sample_rate=self.sample_rate)
+            signals.append(cur_signal)
+        return signals
+
+    def plot(self, output_name, three_d_plot=False, normalize=True):
+        """Plots histograms with the results of the DUET algorithm
+
+        Parameters:
+            output_name (str): path to save plot as
+            three_d_plot (Optional[bool]): Flags whether or not to plot in 3d. Defaults to False
+            normalize (Optional[bool]): Flags whether the matrix should be normalized or not
+        """
+        plt.close('all')
```

Comment:
Same thing here as above, make sure that everything is set correctly before you do anything else.